### ISSUE TITLE
/home/user file problem

### ISSUE THREAD

**Issue by luqtas** [suurjaak/InputScope#27]
i'm using linux and i noticed that this file `InputScope____home__luqtas____local__lib__python3__11__site__packages__inputscope__var__inputscope__db` sometimes needs to be deleted after boot, so the app can work... it's located at my /home directory & it's created again every-time i use the app

**Comment by luqtas** [comment]
ok, actually i'm using 'systemctl reboot/shutdown' and apparently, that's the cause of the problem

**Comment by suurjaak** [comment]
Can you explain a bit more what the problem was? How did systemctl cause problems for InputScope?

**Comment by luqtas** [comment]
i thought that systemctl poweroff/reboot the system "directly" vs. the gnome way of requesting a shutdown/reboot, thus corrupting the file... but i tried again but i could not replicate & also realized there's not much difference between them :D

the only i could find to make it work again is to delete the file

**Comment by suurjaak** [comment]
Was there any error given by the program when it failed?

**Comment by luqtas** [comment]
if python there's any type of "verbose" mode, i can try! because if i try to run `inputscope` in terminal it do nothing, like this:
```
user@pc: ~$ inputscope
user@pc: ~$
```

**Comment by suurjaak** [comment]
There is no verbose mode available, no. 

But you can run the different parts of it individually, e.g.

`inputscope-listener` (or `python -m inputscope.listener`) runs the part that listens to input events and saves them to the database. 

`inputscope-webui` (or `python -m inputscope.webui`) runs the part that serves the web interface.

Perhaps with those you can see whether any errors are encountered.

**Comment by luqtas** [comment]
weird... if i try to execute the -listener/webui it works just fine & afterwards, quitting them with C-c, allows me to execute `inputscope` without it quitting

i can share the file i'm deleting to make it work at the next boot, if that helps

**Comment by suurjaak** [comment]
Yes, seeing the database file would be useful. 

Btw, what is the Python version on your system?

**Comment by luqtas** [comment]
Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]

send you the file through e-mail

**Comment by suurjaak** [comment]
Well, the file I received is certainly not a valid database. It was 5 bytes, just containing "1380" + a null byte.

Are you sure you sent the right file? If you examine it yourself on your computer (for example just with `more`), what do you see?

Also, is the file name truly `InputScope____home__luqtas____local__lib__python3__11__site__packages__inputscope__var__inputscope__db`? And it's just in your user home directory root? Because in reality its path should be `/home/luqtas/.local/lib/python3.11/site-packages/inputscope/var/inputscope.db`. This might actually indicate where the problem is..

**Comment by luqtas** [comment]
"3298" here... but now this one comes from; deleting that file and then starting inputscope again...

and yes, i have the .db file that you mentioned! that one could lead to a better conclusion?

**Comment by suurjaak** [comment]
Oh, now I understand, sorry for the confusion. 

The extra-long-named file in your home directory is a lockfile for ensuring a single instance of InputScope running at a time; it is created by the program main process. It is normal for this file to persist. (Although I should make a change to not pollute the home-directory like that..)

If launching `inputscope` from the command-line exits immediately, then it would imply you already have the program running on your computer. Are you sure something else has not launched it already? This file should not interfere otherwise.

The number inside the lockfile is the process ID of the program instance that created it. You can check your system process list to see what is running under this ID.

**Comment by luqtas** [comment]
could the error be because i'm launching it at startup with Gnome Tweaks? which is strange as i detect this error when i can't open the inputscope webpage...

------- Original Message -------

**Comment by suurjaak** [comment]
So, how exactly are you launching it? Automatically via Gnome Tweaks, or manually from the command-line?

**Comment by luqtas** [comment]
via Gnome Tweaks

**Comment by suurjaak** [comment]
Okay. So is it so that you have set it to launch at startup via Gnome Tweaks, but the program webpage was not coming up? And so you tried to launch it manually, only to run into this issue with this lockfile?

Then I guess the issue is that why is the webpage not working when launched via Gnome Tweaks. Tried it myself, on a Ubuntu host, had no issues. 

What is the result of running `ps -aux | grep inputscope` in terminal?

**Comment by luqtas** [comment]
i had to reboot/poweroff like 3 times to get the error (not launching)...

**ps -aux | grep inputscope** if the file corrupted throws me this,
`luqtas      3018  0.0  0.0   6332  2040 pts/0    S+   15:08   0:00 grep inputscope`


and after deletion of the /home file and starting the app through the same script i use in Gnome Tweaks,
```luqtas      3411  0.0  0.0   2576   900 ?        S    15:12   0:00 /bin/sh /home/luqtas/Downloads/apps/shortcuts/inputscope.sh
luqtas      3413  5.5  0.5 414016 84340 ?        Sl   15:12   0:00 /usr/bin/python3 /home/luqtas/.local/bin/inputscope
luqtas      3440 13.3  0.3 389132 50484 ?        Sl   15:12   0:00 /usr/bin/python3 -m inputscope.listener --quiet
luqtas      3441  2.7  0.3  97904 52656 ?        S    15:12   0:00 /usr/bin/python3 -m inputscope.webui --quiet
luqtas      5086  0.0  0.0   6332  2076 pts/0    S+   15:12   0:00 grep inputscope
```

**Comment by suurjaak** [comment]
Looks like something is rather fragile about the single instance checker under Linux.. sometimes at least. 

If you're willing to test this a bit more, can you edit the file `/home/luqtas/.local/lib/python3.11/site-packages/inputscope/main.py`, and on the last page replace

```python
        MainApp(redirect=True).MainLoop() # stdout/stderr directed to wx popup
```

with

```python
        try: MainApp(redirect=True).MainLoop() # stdout/stderr directed to wx popup
        finally: del singlechecker
```
Does this help with cleaning up the lockfile properly?

**Comment by luqtas** [comment]
nope, it doesn't change anything and just for the sake of knowing if i'm doing it right; it stays like this, alright?

```python
def main():
    """Program entry point."""
    if conf.Frozen: multiprocessing.freeze_support()
    conf.init(), db.init(conf.DbPath, conf.DbStatements)
    try: db.execute("PRAGMA journal_mode = WAL")
    except Exception: pass

    if wx:
        name = re.sub(r"\W", "__", "-".join([conf.Title, conf.DbPath]))
        singlechecker = wx.SingleInstanceChecker(name)
        if singlechecker.IsAnotherRunning(): sys.exit()

        #MainApp(redirect=True).MainLoop() # stdout/stderr directed to wx popup
        try: MainApp(redirect=True).MainLoop() # stdout/stderr directed to wx popup
        finally: del singlechecker
    else:
        model = Model()
        if tk:
            widget = tk.Tk() # Use Tkinter instead to get screen size
            size = [0, 0, widget.winfo_screenwidth(), widget.winfo_screenheight()]
            model.log_resolution([size])
            widget.destroy()
        print("wxPython not available, using basic command line interface.")
        print("Web interface running at %s" % conf.WebUrl)
        try:
            model.run()
        except IOError as e:
            if e.errno != errno.EINTR: raise # Interrupted syscall, probably sleep
        except KeyboardInterrupt:
            model.stop()
```

**Comment by suurjaak** [comment]
Yes, your changes are all correct. 

I will try making a custom implementation of the single instance checker, will get back to you with it, hopefully in a day or two. 

In the meanwhile, you can undo these changes and just comment out the lines regarding `singlechecker`. As long as you don't launch the application multiple times, it should work fine.

**Comment by luqtas** [comment]
thanks!

i like your app so much that for now, when i reboot/shutdown, it deletes the /home file before doing it :P
-------- Original Message --------

**Comment by suurjaak** [comment]
Can you try downloading and installing this version: [InputScope-1.10.dev6.tar.gz](https://github.com/suurjaak/InputScope/files/13060666/InputScope-1.10.dev6.tar.gz). I changed the single instance check logic to be more robust in Linux. 

It can be installed via `pip install InputScope-1.10.dev6.tar.gz`.

Note that installing it will overwrite your current input history, so if you want to retain it, make a backup of the directory `.local/lib/python3.11/site-packages/inputscope/var/` before installing, and move it back afterwards. 

(And the lockfile will now go under `.local/share/inputscope/`).

**Comment by luqtas** [comment]
ok: day 2 of trying to replicate the error and i couldn't... so your fix resolved the problem, i think! although now i can not access the mouse statics! it throws me this error when clicking at (moves, clicks or scrolls);
```
Error: 500 Internal Server Error
Sorry, the requested URL 'http://localhost:8099/mouse/clicks' caused an error:
Internal Server Error
```
the .config file is set to register the mouse...

**Comment by suurjaak** [comment]
Very good that the lockfile error is resolved. I will make an official release for it soon. 

Regarding the new error: can you shut down the auto-launched InputScope, and launch `inputscope` or `inputscope-webui` manually in a terminal? It should print the error to console in more detail. 

To shut down the program running in the background, you would need to terminate its main process. You can look up the process ID via `ps -aux | grep inputscope`, and terminate it via `kill -2 THATPID` (the relevant process should be listed as something like `/usr/bin/python3 /home/luqtas/.local/bin/inputscope`).

**Comment by luqtas** [comment]
```
Traceback (most recent call last):
  File "/home/luqtas/.local/lib/python3.11/site-packages/bottle.py", line 876, in _handle
    return route.call(**args)
           ^^^^^^^^^^^^^^^^^^
  File "/home/luqtas/.local/lib/python3.11/site-packages/bottle.py", line 1759, in wrapper
    rv = callback(*a, **ka)
         ^^^^^^^^^^^^^^^^^^
  File "/home/luqtas/.local/lib/python3.11/site-packages/inputscope/webui.py", line 166, in inputdetail
    stats_texts, app_stats, heatmap_stats, events = stats_mouse(events, table, count)
                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/luqtas/.local/lib/python3.11/site-packages/inputscope/webui.py", line 308, in stats_mouse
    for row in db.fetch("screen_sizes", order="dt"):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/luqtas/.local/lib/python3.11/site-packages/inputscope/db.py", line 28, in fetch
    return select(table, cols, where, group, order, limit, **kwargs).fetchall()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/sqlite3/dbapi2.py", line 65, in convert_timestamp
    datepart, timepart = val.split(b" ")
    ^^^^^^^^^^^^^^^^^^
ValueError: not enough values to unpack (expected 2, got 1)
```

**Comment by suurjaak** [comment]
Well, that was unexpected.

Can you check one thing: in a terminal, launch `python3`, and in the console there execute the following, one by one:

```python
import sqlite3

db = sqlite3.connect("/home/luqtas/.local/lib/python3.11/site-packages/inputscope/var/inputscope.db")

for row in db.execute("SELECT * FROM screen_sizes"): print(row)
```
What are the results?

**Comment by luqtas** [comment]
```
... 
(1, 'keys', '2023-10-07', 187, None, None, None)
(2, 'combos', '2023-10-07', 13, None, None, None)
(3, 'moves', '2023-10-07', 703, None, None, None)
(4, 'clicks', '2023-10-07', 295, None, None, None)
(5, 'scrolls', '2023-10-07', 10, None, None, None)
(6, 'moves', '2023-10-08', 18625, None, None, None)
(7, 'keys', '2023-10-08', 16585, None, None, None)
(8, 'combos', '2023-10-08', 2189, None, None, None)
(9, 'clicks', '2023-10-08', 2446, None, None, None)
(10, 'scrolls', '2023-10-08', 283, None, None, None)
(11, 'moves', '2023-10-09', 14444, None, None, None)
(12, 'clicks', '2023-10-09', 1832, None, None, None)
(13, 'keys', '2023-10-09', 2803, None, None, None)
(14, 'combos', '2023-10-09', 220, None, None, None)
(15, 'scrolls', '2023-10-09', 1144, None, None, None)
(16, 'clicks', '2023-10-10', 2453, None, None, None)
(17, 'moves', '2023-10-10', 15742, None, None, None)
(18, 'keys', '2023-10-10', 10506, None, None, None)
(19, 'scrolls', '2023-10-10', 275, None, None, None)
(20, 'combos', '2023-10-10', 1257, None, None, None)
(21, 'moves', '2023-10-11', 4602, None, None, None)
(22, 'clicks', '2023-10-11', 1016, None, None, None)
(23, 'keys', '2023-10-11', 9982, None, None, None)
(24, 'combos', '2023-10-11', 1668, None, None, None)
(25, 'scrolls', '2023-10-11', 278, None, None, None)
(26, 'clicks', '2023-10-12', 4210, None, None, None)
(27, 'moves', '2023-10-12', 17201, None, None, None)
(28, 'scrolls', '2023-10-12', 318, None, None, None)
(29, 'keys', '2023-10-12', 15167, None, None, None)
(30, 'combos', '2023-10-12', 1533, None, None, None)
(31, 'moves', '2023-10-13', 16297, None, None, None)
(32, 'clicks', '2023-10-13', 2554, None, None, None)
(33, 'keys', '2023-10-13', 14215, None, None, None)
(34, 'combos', '2023-10-13', 2854, None, None, None)
(35, 'scrolls', '2023-10-13', 98, None, None, None)
(36, 'clicks', '2023-10-14', 3086, None, None, None)
(37, 'moves', '2023-10-14', 13636, None, None, None)
(38, 'keys', '2023-10-14', 4361, None, None, None)
(39, 'combos', '2023-10-14', 363, None, None, None)
(40, 'scrolls', '2023-10-14', 464, None, None, None)
(41, 'moves', '2023-10-15', 22554, None, None, None)
(42, 'clicks', '2023-10-15', 3993, None, None, None)
(43, 'scrolls', '2023-10-15', 269, None, None, None)
(44, 'keys', '2023-10-15', 16400, None, None, None)
(45, 'combos', '2023-10-15', 2114, None, None, None)
(46, 'clicks', '2023-10-16', 2288, None, None, None)
(47, 'moves', '2023-10-16', 11971, None, None, None)
(48, 'scrolls', '2023-10-16', 219, None, None, None)
(49, 'keys', '2023-10-16', 14629, None, None, None)
(50, 'combos', '2023-10-16', 6150, None, None, None)
(51, 'keys', '2023-10-17', 14825, None, None, None)
(52, 'combos', '2023-10-17', 10104, None, None, None)
(53, 'moves', '2023-10-17', 13963, None, None, None)
(54, 'clicks', '2023-10-17', 2312, None, None, None)
(55, 'scrolls', '2023-10-17', 139, None, None, None)
(56, 'keys', '2023-10-18', 14630, None, None, None)
(57, 'combos', '2023-10-18', 4013, None, None, None)
(58, 'moves', '2023-10-18', 23179, None, None, None)
(59, 'clicks', '2023-10-18', 4097, None, None, None)
(60, 'scrolls', '2023-10-18', 925, None, None, None)
(61, 'keys', '2023-10-19', 12514, None, None, None)
(62, 'moves', '2023-10-19', 38106, None, None, None)
(63, 'clicks', '2023-10-19', 5972, None, None, None)
(64, 'combos', '2023-10-19', 725, None, None, None)
(65, 'scrolls', '2023-10-19', 448, None, None, None)
(66, 'keys', '2023-10-20', 7550, None, None, None)
(67, 'combos', '2023-10-20', 677, None, None, None)
(68, 'moves', '2023-10-20', 9681, None, None, None)
(69, 'clicks', '2023-10-20', 1640, None, None, None)
(70, 'scrolls', '2023-10-20', 120, None, None, None)
(71, 'keys', '2023-10-21', 2234, None, None, None)
(72, 'combos', '2023-10-21', 183, None, None, None)
(73, 'moves', '2023-10-21', 5213, None, None, None)
(74, 'clicks', '2023-10-21', 1052, None, None, None)
(75, 'scrolls', '2023-10-21', 178, None, None, None)
```

**Comment by suurjaak** [comment]
Very bizarre. That is completely wrong content for this table, it looks like data from the `counts`-table instead, which holds daily event counts. Something has gone terribly wrong somewhere.. 

Can you perhaps provide me with the database? 

To make the database size smaller before sending, and also to remove personal information, you can do the following:
- make a copy of the database file
- in a terminal, launch `python3`, and then execute the following, one by one:
```python

import sqlite3

db = sqlite3.connect("PUT THE PATH TO THE COPY HERE", isolation_level=None)

for table in ("moves", "clicks", "scrolls", "keys", "combos"): db.execute("DELETE FROM " + table)

db.execute("VACUUM")

db.close()
```

**Comment by luqtas** [comment]
it seems that Github doesn't allow me to drag & drop a .db file and i got an error doing what you typed!
```
Python 3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sqlite3
>>> db = sqlite3.connect("PUT THE PATH TO THE COPY HERE", isolation_level=None)
>>> for table in ("moves", "clicks", "scrolls", "keys", "combos"): db.execute("DELETE FROM " + table)
... db.execute("VACUUM")
  File "<stdin>", line 2
    db.execute("VACUUM")
    ^^
SyntaxError: invalid syntax
>>> db.close()
```
did i corrupted the file?

**Comment by suurjaak** [comment]
The syntax error is from not executing the statements one by one, so the last two statements got combined into a single, invalid one. Simply press Enter twice after each statement. 

Also, you need to replace the `PUT THE PATH TO THE COPY HERE` part with the actual path of the copied file.

You can send the file to my e-mail.

**Comment by luqtas** [comment]
weird... i decided to delete the file (.../var/inputscope.db) & running inputscope now, the mouse section just works! could it be because the update i did with `pip install .`?

**Comment by suurjaak** [comment]
Probably something got corrupted in the database file, content mixed up.. somehow.. and that caused the web page to not work. Now that you have deleted the old corrupted database file, all is as it should be. 

A bit pity that I did not get to see that corrupted file, could have given some inkling of what went wrong. But no matter. 

I will make an official public release with the lockfile change soon. Keeping this GitHub issue open until then.

**Comment by suurjaak** [comment]
The newly released version 1.10 should fix this issue.

Apologies for the delay.
